src/ewah.zig

```diff
@@ -0,0 +1,251 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const mem = std.mem;
+
+/// Encode or decode a bitset using Daniel Lemire's EWAH codec.
+/// ("Histogram-Aware Sorting for Enhanced Word-Aligned Compression in Bitmap Indexes")
+///
+/// EWAH uses only two types of words, where the first type is a 64-bit verbatim ("literal") word.
+/// The second type of word is a marker word:
+/// * The first bit indicates which uniform word will follow.
+/// * The next 31 bits are used to store the number of uniform words.
+/// * The last 32 bits are used to store the number of literal words following the uniform words.
+/// EWAH bitmaps begin with a marker word. A 'marker' looks like:
+///
+///     [uniform_bit:u1][uniform_word_count:u31(LE)][literal_word_count:u32(LE)]
+///
+/// and is immediately followed by `literal_word_count` 64-bit literals.
+/// When decoding a marker, the uniform words precede the literal words.
+///
+/// This encoding requires that the architecture is little-endian with 64-bit words.
+pub fn EWAH(comptime Word: type) type {
+    const word_bits = @bitSizeOf(Word);
+
+    const marker_uniform_word_count_max = (1 << ((word_bits / 2) - 1)) - 1;
+    const marker_literal_word_count_max = (1 << (word_bits / 2)) - 1;
+
+    const MarkerUniformCount = std.math.IntFittingRange(0, marker_uniform_word_count_max); // Word=usize → u31
+    const MarkerLiteralCount = std.math.IntFittingRange(0, marker_literal_word_count_max); // Word=usize → u32
+
+    const Marker = packed struct {
+        uniform_bit: u1, // Whether the uniform word is all 0s or all 1s.
+        uniform_word_count: MarkerUniformCount, // 31-bit number of uniform words following the marker.
+        literal_word_count: MarkerLiteralCount, // 32-bit number of literal words following the uniform words.
+    };
+
+    comptime {
+        assert(@import("builtin").target.cpu.arch.endian() == std.builtin.Endian.Little);
+        assert(@typeInfo(Word).Int.signedness == .unsigned);
+        assert(word_bits % 8 == 0); // A multiple of a byte, so that words can be cast to bytes.
+        assert(@bitSizeOf(Marker) == word_bits);
+        assert(@sizeOf(Marker) == @sizeOf(Word));
+
+        assert(@bitSizeOf(MarkerUniformCount) % 2 == 1);
+        assert(std.math.maxInt(MarkerUniformCount) == marker_uniform_word_count_max);
+
+        assert(@bitSizeOf(MarkerLiteralCount) % 2 == 0);
+        assert(std.math.maxInt(MarkerLiteralCount) == marker_literal_word_count_max);
+    }
+
+    return struct {
+        const Self = @This();
+
+        inline fn marker_word(mark: Marker) Word {
+            return @bitCast(Word, mark);
+        }
+
+        /// Decodes the compressed bitset in `source` into `set`. Panics if `source`'s encoding is invalid.
+        /// Returns the number of *words* written to `target_words`.
+        pub fn decode(source: []align(@alignOf(Word)) const u8, target_words: []Word) usize {
+            assert(source.len % @sizeOf(Word) == 0);
```

Comment:
> If I understand correctly, I think at present we will actually fall all the way through and return 0 if `source.len` is 0?

Yes. Which, in the context of the codec, is correct. Though in our particular use case (`BlockFreeSet`) we won't ever have an empty bitset.

> Do we have any expectations around the length of `target_words`?

There is one at the end of decoding (though its mostly just for clarity -- if it were actually violated, the `mem.set`/`mem.copy` would have already failed). But beforehand we make the assumption that the bitset's size is being stored out-of-band of the codec. And the caller _is_ allowed to pass a too-large slice -- that's why `decode` returns the number of words written to the target.

It would be nice to assert that the `source`/`target` slice memory doesn't overlap (i.e. they don't alias). I see that zig has a `noalias` keyword, but it isn't documented.

Reply:
> Yes. Which, in the context of the codec, is correct. Though in our particular use case (BlockFreeSet) we won't ever have an empty bitset.

Surely for EWAH though, we always expect an encoded source to contain at least one word, being the marker? If the stream is empty, then we should see source containing a single marker with a `uniform_word_count` of 0 and a `literal_word_count` of 0. In other words, we should assert that `source.len` can never be 0 and must always be at least a word.

Reply:
> And the caller is allowed to pass a too-large slice -- that's why decode returns the number of words written to the target.

100%.

> It would be nice to assert that the source/target slice memory doesn't overlap (i.e. they don't alias).

Great idea. We do something a very slightly similar in `Journal` where we do:
```zig
// Memory must not be owned by `self.headers` as these may be modified concurrently:
assert(@ptrToInt(buffer.ptr) < @ptrToInt(self.headers.ptr) or
    @ptrToInt(buffer.ptr) > @ptrToInt(self.headers.ptr) + self.size_headers);
```

Reply:
Why isn't it

```
assert(@ptrToInt(buffer.ptr) + buffer.len < @ptrToInt(self.headers.ptr) or
    @ptrToInt(buffer.ptr) > @ptrToInt(self.headers.ptr) + self.size_headers);
```

?

Reply:
or

```
assert(@ptrToInt(buffer.ptr) + buffer.len <= @ptrToInt(self.headers.ptr) or
    @ptrToInt(buffer.ptr) >= @ptrToInt(self.headers.ptr) + self.size_headers);
```

rather. And the lengths would need to be adjusted to bytes (though in this case they are already).

Reply:
Yes, that would be better and stronger (although I think it would need to use `<=` and `>=` in case the buffers are exactly directly adjacent).

The intention is just to ensure that `buffer.ptr` is before or after the `self.headers` range, to catch any accidental reuse of something derived from `self.headers` (which is fairly easy to do by mistake when looking things up in the journal).

Reply:
Ah, we're commenting concurrently! Good catch! 😄